homeassistant/helpers/script.py

```diff
@@ -626,74 +642,47 @@ async def _async_wait_template_step(self):
             self._variables["wait"]["completed"] = True
             return
 
+        futures, timeout_handle, timeout_future = self._async_futures_with_timeout(
+            timeout
+        )
+        done = self._hass.loop.create_future()
+        futures.append(done)
+
         @callback
         def async_script_wait(entity_id, from_s, to_s):
             """Handle script after template condition is true."""
-            # pylint: disable=protected-access
-            wait_var = self._variables["wait"]
-            if to_context and to_context._when:
-                wait_var["remaining"] = to_context._when - self._hass.loop.time()
-            else:
-                wait_var["remaining"] = timeout
-            wait_var["completed"] = True
-            done.set()
+            self._async_set_remaining_time_var(timeout_handle)
+            self._variables["wait"]["completed"] = True
+            _set_result_unless_done(done)
 
-        to_context = None
         unsub = async_track_template(
             self._hass, wait_template, async_script_wait, self._variables
         )
-
         self._changed()
-        done = asyncio.Event()
```

Comment:
wait template and wait trigger were almost the same code so they can share waiting with `_async_wait_with_optional_timeout`